# Домашнее задание №2

**

## Часть 1. ML workflow (**всего 5 баллов**)

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import mean_squared_error, r2_score

from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

from itertools import product

In [2]:
import warnings
warnings.filterwarnings('ignore')

### Загрузим данные для работы. 

In [3]:
df = pd.read_csv("winequality-red.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [4]:
df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

Будем решать задачу регрессии: необходимо предсказать качество вина на основе его характеристик

### Шаг 1.  (**0.2 балла**)
Создайте матрицу X объект-признак и целевой вектор y ("quality")

In [5]:
X = df.iloc[:,:-1]

y = df['quality']

In [6]:
X

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  
0         9.4  
1         9.8  
2         9.8  
3         9.8  
4         9.4  
...       ...  
1594     10.5  
1595     11.2  
1596     11.0  
1597     10.2  
1598     11.0  

[1599 rows x 11 columns]

In [7]:
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

### Шаг 2. (**0.2 балла**)
Разбейте данные на train и test (доля тестовых данных - 30%).

In [8]:
# Разобьем данные на обучающую и тестовую выборки

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=668)

### Шаг 3. (**0.2 балла**)
Обучите линейную регрессию на тренировочных данных и сделайте предсказания на train и на test.

In [9]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

lr = LinearRegression()
# объявляем модель
lr.fit(X_train,y_train)
# обучаем модель на тренировочных данных

pred_train = lr.predict(X_train)
#делаем предсказания на трейне 
pred_test = lr.predict(X_test)
#делаем предсказания на тесте 

 #отдельно делаем предсказание, чтобы посмотреть, не переобучиллась ли модель

### Шаг 4. (**0.4 балла**)
Выведите на экран ошибку MSE на train и на test, затем выведите на экран ошибку r2 на train и test.  

In [10]:
print("Train MSE: {}".format(mean_squared_error(y_train, pred_train)))
#считаем квадратичную ошибку - сравниваем
print("Test MSE: {}".format(mean_squared_error(y_test, pred_test)))

#на тесте ошибка меньше ( на трейне большая ошибка)- модель недообучилась

Train MSE: 0.42259758262551556
Test MSE: 0.4077004129367436


In [11]:
print("R^2_train: {}".format(lr.score(X_train, y_train)))
print("R^2_test: {}".format(lr.score(X_test, y_test)))

R^2_train: 0.3651286255201631
R^2_test: 0.34034989367538104


Другой способ нахождения R^2:

In [12]:
from sklearn.metrics import r2_score

y_pred_test = lr.predict(X_test)
y_pred_train = lr.predict(X_train)

print("R^2_train: {}".format(r2_score(y_train, y_pred_train)))
      
print("R^2_test: {}".format(r2_score(y_test, y_pred_test)))

R^2_train: 0.3651286255201631
R^2_test: 0.34034989367538104


In [13]:
lr.coef_

array([ 6.61033370e-02, -1.12798358e+00, -2.26332715e-01,  3.14128318e-02,
       -1.74047331e+00,  6.53564104e-03, -3.62435934e-03, -5.13627297e+01,
       -1.74055405e-01,  9.86364783e-01,  2.49439404e-01])

### Шаг 5. (**0.5 балла**)
Вычислите среднее качество (r2) модели на кросс-валидации с k=5 фолдами.

In [14]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_validate 

kf = KFold(n_splits=5, random_state=None, shuffle=True)  #фиксирует разбиение на блоки  
kf.get_n_splits(X) #разбили данные на блоки 

5

In [15]:
# проверим на кросс-валидации значения ошибок MSE 
# для линейной регрессии, обученной с помощью функционала MSE

# cv_res = cross_validate(lr, # модель
#                      X, # данные, на которых обучаем 
#                      y,# данные, на которых обучаем 
#                      scoring='neg_mean_squared_error', # метрика, которую нужно оценить
#                         #(по умолчанию максимизируется, а нам нужно mse минимизировать - neg_mse )
#                      cv=kf # количество разбиений или класс-сплиттер 
#                         #(cv- количество фолдов, а ранее мы зафиксировали разбиение - используем)
#                     )

# print(f"test mse errors are {cv_res['test_score']}") #посмотрим результаты на тесте 
# print(f"mean test mse = {-cv_res['test_score'].mean()}") #смотрим на среднее качество - получаем менее смещенную оценку 

In [16]:
# проверим на кросс-валидации значения ошибок R2 
# для линейной регрессии, обученной с помощью функционала MSE

cv_res = cross_validate(lr, # модель
                     X, # данные, на которых обучаем 
                     y,# данные, на которых обучаем 
                     scoring='r2', # метрика, которую нужно оценить
                        #(по умолчанию максимизируется)
                     cv=kf # количество разбиений или класс-сплиттер 
                        #(cv- количество фолдов, а ранее мы зафиксировали разбиение - используем)
                    )

print(f"test r2 are {cv_res['test_score']}") #посмотрим результаты на тесте 
print(f"mean test r2 = {cv_res['test_score'].mean()}") #смотрим на среднее качество - получаем менее смещенную оценку 

test r2 are [0.41617274 0.31694518 0.31963687 0.26798769 0.3496642 ]
mean test r2 = 0.3340813346720054


### Шаг 6.  (**0.5 балла**)
Теперь примените линейную регрессию с L1-регуляризацией (Lasso) для данной задачи. Объявите модель и подберите параметр регуляризации alpha по сетке. Ищите alpha в диапазоне (0.1, 1.1) с шагом 0.1. 

Осуществите подбор параметра alpha по тренировочным данным (Xtrain, ytrain).

In [17]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Lasso

grid_search = GridSearchCV(Lasso(), {'alpha': np.arange(0.1, 1.1, 0.1)}, cv=5)
grid_search.fit(X_train, y_train)
best_alpha = grid_search.best_params_['alpha']

print(f'Optimal alpha value is {best_alpha}')

Optimal alpha value is 0.1


### Шаг 7.  (**0.5 балла**)
Выведите наилучший алгоритм и наилучшее качество по результатам подбора alpha (best_estimator_ и best_score_).

In [18]:
grid_search.best_estimator_, grid_search.best_score_

(Lasso(alpha=0.1), 0.2284940602915261)

### Шаг 8.  (**0.5 балла**)

С помощью найденного best_estimator_ сделайте предсказание на тестовых данных и выведите на экран r2-score на тесте.

In [19]:
y_pred2 = grid_search.best_estimator_.predict(X_test)

print("R^2_test: {}".format(r2_score(y_test,y_pred2)))

R^2_test: 0.2506119423956005


### Шаг 9.  (**0.5 балла**)

Попробуем улучшить качество модели за счет добавления полиномиальных признаков. Создайте pipeline, состоящий из добавления полиномиальных признаков степени 2, а затем применения линейной регрессии.

Затем вычислите r2-score этой модели на кросс валидации с пятью фолдами.

In [20]:
#избегаем повторение 
import copy

cols = copy.deepcopy(X.columns) #сохраним список названий 
print(cols)

for i in range(len(cols)):
    for j in range(i,len(cols)):
        col1=cols[i]
        col2=cols[j]
        
        col_name = col1 + '_x_' + col2
        if col_name not in X.columns:
            X[col_name] = X[col1]*X[col2]#два одинаковые признака  x1x2=x2x1
X 

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol'],
      dtype='object')


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      ...  density_x_density  density_x_pH  density_x_sulphates  \
0     ...           0.995605      3.502278             0.558768   
1     ...           0.993610      3.189760             0.677824   
2     ...           0.994009      3.250220             0.648050   
3     ...           0.996004      3.153680             0.578840   
4     ...           0.995605      3.502278             0.558768   
...   ...                ...           ...                  ...   
1594  ...           0.989826      3.432405             0.577042   
1595  ...           0.990264      3.502822             0.756291   
1596  ...           0.991498      3.405431             0.746805   
1597  ...           0.990961      3.553828             0.706784   
1598  ...           0.991000      3.374711             0.657023   

      density_x_alcohol  pH_x_pH  pH_x_sulphates  pH_x_alcohol  \
0              9.379320  12.3201          1.9656        32.994   
1              9.768640  10.2400          2.1760        31.360   
2              9.770600  10.6276          2.1190        31.948   
3              9.780400   9.9856          1.8328        30.968   
4              9.379320  12.3201          1.9656        32.994   
...                 ...      ...             ...           ...   
1594          10.446450  11.9025          2.0010        36.225   
1595          11.145344  12.3904          2.6752        39.424   
1596          10.953140  11.6964          2.5650        37.620   
1597          10.153794  12.7449          2.5347        36.414   
1598          10.950390  11.4921          2.2374        37.290   

      sulphates_x_sulphates  sulphates_x_alcohol  alcohol_x_alcohol  
0                    0.3136                5.264              88.36  
1                    0.4624                6.664              96.04  
2                    0.4225                6.370              96.04  
3                    0.3364                5.684              96.04  
4                    0.3136                5.264              88.36  
...                     ...                  ...                ...  
1594                

In [21]:
X.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'fixed acidity_x_fixed acidity',
       'fixed acidity_x_volatile acidity', 'fixed acidity_x_citric acid',
       'fixed acidity_x_residual sugar', 'fixed acidity_x_chlorides',
       'fixed acidity_x_free sulfur dioxide',
       'fixed acidity_x_total sulfur dioxide', 'fixed acidity_x_density',
       'fixed acidity_x_pH', 'fixed acidity_x_sulphates',
       'fixed acidity_x_alcohol', 'volatile acidity_x_volatile acidity',
       'volatile acidity_x_citric acid', 'volatile acidity_x_residual sugar',
       'volatile acidity_x_chlorides',
       'volatile acidity_x_free sulfur dioxide',
       'volatile acidity_x_total sulfur dioxide', 'volatile acidity_x_density',
       'volatile acidity_x_pH', 'volatile acidity_x_sulphates',
       'volatile acidity_x_alcohol', 'citric acid_x_citric acid',
    

In [22]:
# проверим на кросс-валидации значения ошибок R2 
# для линейной регрессии, обученной с помощью функционала MSE
from sklearn.model_selection import cross_validate 
cv_res = cross_validate(lr, # модель
                     X, # данные, на которых обучаем 
                     y,# данные, на которых обучаем 
                     scoring='r2', # метрика, которую нужно оценить
                        #(по умолчанию максимизируется)
                     cv=5 # количество разбиений или класс-сплиттер 
                    )

print(f"test r2 are {cv_res['test_score']}") #посмотрим результаты на тесте 
print(f"mean test r2 = {cv_res['test_score'].mean()}") #смотрим на среднее 

test r2 are [-0.0399065   0.29930017  0.38080623  0.30145898  0.20882197]
mean test r2 = 0.23009617074676797


### Шаг 10.  (**0.5 балла**)
Обучите модель (pipeline) на тренировочных данных и сделайте предсказания для train и test, затем выведите на экран r2-score и MSE на тренировочных и на тестовых данных.

In [23]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

X, y = make_classification(random_state=0)
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                     random_state=668)

pipe = Pipeline([('scaler', StandardScaler()), ('svc', SVC())])
pipe.fit(X_train, y_train)

y_pred_test = pipe.predict(X_test)
y_pred_train = pipe.predict(X_train)

print("R^2_train: {}".format(pipe.score(X_train, y_train)))
print("R^2_test: {}".format(pipe.score(X_test, y_test)))

print("")
print("Train MSE: {}".format(mean_squared_error(y_train, y_pred_train)))
print("Test MSE: {}".format(mean_squared_error(y_test, y_pred_test)))

R^2_train: 0.9733333333333334
R^2_test: 0.8

Train MSE: 0.02666666666666667
Test MSE: 0.2


In [24]:
pipe.score(X_test, y_test)

0.8

### Сделайте выводы. Для этого ответьте на вопросы: (**1 балл**)

**1) Хорошее ли качество показала исходная модель (линейная регрессия без регуляризации)? Является ли эта модель переобученной?**

Качество исзодной модели 

$MSE_{tain}: 0.42259758262551556$

$MSE_{test}: 0.4077004129367436$

Мы не наблюдаем большую разницу в качестве на тренировочных и тестовых данных (модель не подгоняется под тренированные данные), на test ошибка меньше,чем на train.
Также значения параметров (весов) модели небольшое, обратное могло свидетельствовать о переобучении модели. 

$R^2_{train}: 0.3651286255201631$

$R^2_{test}: 0.34034989367538104$

Качество модели согласно значению $R^2$ довольно низкое, однако модель не является переобученной

In [25]:
lr.coef_# проверка весов, не слишком ли большие 

array([ 6.61033370e-02, -1.12798358e+00, -2.26332715e-01,  3.14128318e-02,
       -1.74047331e+00,  6.53564104e-03, -3.62435934e-03, -5.13627297e+01,
       -1.74055405e-01,  9.86364783e-01,  2.49439404e-01])

**2)Помогла ли L1-регуляризация улучшить качество модели?**

$𝑅^2_{first}:0.34034989367538104$

$R^2_{L1}: 0.2506119423956005$

Нет, качество модели ухудшилось (отбор признаков привел к занулению признаков, в связи с чем модель стала разреженной)

**3) Помогло ли добавление полиномов второй степени улучшить качество модели? Как добавление новых признаков повлияло на переобучение?**

Ранее $R^2 = 0.348525835304305$

После добавления полиномов второй степени $R^2=0.23009617074676797$

*при обучении модели pipline - $R^2 = 0.8$

Наблюдаем избыточную сложность модели, обусловленную большим количеством признаков (весов). Лишние степени свободы в модели "тратятся" на чрезмерно точную подгонку под обучащую выборку.

Теперь модель не находит закономерность, а просто подстраивается к данным (на тестовой выборке наблюдаем большую ошибку)+ $R^2$ близок к $1$

$MSE_{train}: 0.02666666666666667$

$MSE_{test}: 0.2$

Замечаем - ошибка на тесте намного больше ошибки на трейне

### *Попытайтесь улучшить модель (добейтесь наилучшего качества) - можно использовать любые методы.

При улучшении качества r2 на 0.1-0.2 +1 балл, при большем улучшении +2 балла (дополнительно к 5 баллам за основную часть).

In [26]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge

grid_search = GridSearchCV(Ridge(), {'alpha': np.arange(0, 1.1, 0.1)}, cv=5)
grid_search.fit(X_train, y_train)
best_alpha = grid_search.best_params_['alpha']

print(f'Optimal alpha value is {best_alpha}')

Optimal alpha value is 1.0


In [27]:
grid_search.best_score_

0.36030747424059933

## Часть 2. Target encoding (**всего 5 баллов**)

В этом части домашнего задания вы будете работать с выборкой `1C`. Вам нужно посчитать счетчики для `item_id` четырьмя способами:

    1) При помощи KFold схемы;  
    2) При помощи Leave-one-out схемы;
    3) При помощи smoothing схемы;
    4) При помощи expanding mean схемы.

### Подготовка данных

In [29]:
sales = pd.read_csv('sales_train_v2.csv') #пришлось заменить название файла
sales.columns = ['date', 'date_block_num', 'shop_id', 'item_id', 'item_price', 'target']
sales

date  date_block_num  shop_id  item_id  item_price  target
0        02.01.2013               0       59    22154      999.00     1.0
1        03.01.2013               0       25     2552      899.00     1.0
2        05.01.2013               0       25     2552      899.00    -1.0
3        06.01.2013               0       25     2554     1709.05     1.0
4        15.01.2013               0       25     2555     1099.00     1.0
...             ...             ...      ...      ...         ...     ...
2935844  10.10.2015              33       25     7409      299.00     1.0
2935845  09.10.2015              33       25     7460      299.00     1.0
2935846  14.10.2015              33       25     7459      349.00     1.0
2935847  22.10.2015              33       25     7440      299.00     1.0
2935848  03.10.2015              33       25     7460      299.00     1.0

[2935849 rows x 6 columns]

In [30]:
index_cols = ['shop_id', 'item_id', 'date_block_num']

# For every month we create a grid from all shops/items combinations from that month
grid = [] 
for block_num in sales['date_block_num'].unique():
    cur_shops = sales[sales['date_block_num']==block_num]['shop_id'].unique()
    cur_items = sales[sales['date_block_num']==block_num]['item_id'].unique()
    grid.append(np.array(list(product(*[cur_shops, cur_items, [block_num]])),dtype='int32'))

#turn the grid into pandas dataframe
grid = pd.DataFrame(np.vstack(grid), columns = index_cols,dtype=np.int32)

#get aggregated values for (shop_id, item_id, month)
gb = sales.groupby(index_cols,as_index=False).agg({'target':'sum'})

#join aggregated data to the grid
all_data = pd.merge(grid,gb,how='left',on=index_cols).fillna(0)
#sort the data
all_data.sort_values(['date_block_num','shop_id','item_id'],inplace=True)

In [31]:
all_data

shop_id  item_id  date_block_num  target
139255          0       19               0     0.0
141495          0       27               0     0.0
144968          0       28               0     0.0
142661          0       29               0     0.0
138947          0       32               0     6.0
...           ...      ...             ...     ...
10768834       59    22162              33     0.0
10769024       59    22163              33     0.0
10769690       59    22164              33     0.0
10771216       59    22166              33     0.0
10770511       59    22167              33     0.0

[10913850 rows x 4 columns]

### Mean encodings без регуляризации

После проделанной технической работы, мы готовы посчитать счетчики для переменной `item_id`. 

Ниже приведены две реализации подсчета счетчиков без регуляризации. Можно использовать данный код в качестве стартовой точки для реализации различных техник регуляризации.

#### Способ 1

In [32]:
# Calculate a mapping: {item_id: target_mean}
item_id_target_mean = all_data.groupby('item_id').target.mean()

# In our non-regularized case we just *map* the computed means to the `item_id`'s
all_data['item_target_enc'] = all_data['item_id'].map(item_id_target_mean)

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


#### Способ 2

In [33]:
'''
     Differently to `.target.mean()` function `transform` 
   will return a dataframe with an index like in `all_data`.
   Basically this single line of code is equivalent to the first two lines from of Method 1.
'''
all_data['item_target_enc'] = all_data.groupby('item_id')['target'].transform('mean')

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


###  KFold схема (**1.25 балла**)

Необходимо реализовать Kfold схему с пятью фолдами. Используйте KFold(5) из sklearn.model_selection. 

1. Разбейте данные на 5 фолдов при помощи `sklearn.model_selection.KFold` с параметром `shuffle=False`.
2. Проитерируйтесь по фолдам: используйте 4 обучающих фолда для подсчета средних значений таргета по `item_id` и заполните этими значениями валидационный фолд на каждой итерации.

Обратите внимание на **Способ 1** из примера. В частности, изучите, как работают функции map и pd.Series.map. Они довольно полезны во многих ситуациях. 

In [34]:
kf = KFold(n_splits=5, random_state=None, shuffle=False)  #фиксирует разбиение на блоки  
folds = kf.split(all_data) #разбили данные на блоки 

for train_fold, test_fold in folds:
    fold_data = all_data.iloc[train_fold]
    
    item_id_target_mean = fold_data.groupby('item_id').target.mean()
    all_data['item_target_enc'].iloc[test_fold] = (all_data.iloc[test_fold]['item_id']
                                                           .map(item_id_target_mean))
    
# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values  
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.41645907127988024


Ожидаемый ответ 0.4165

### Leave-one-out схема (**1.25 балла**)

Необходимо реализовать leave-one-out схему . Учтите, если вы запустите код из первого задания, задав количество фолдов такое же как размер выборки, то вы, вероятно, получите правильный ответ, но ждать будете очень-очень долго.

Для более быстрой реализации подсчета среднего таргета на всех объектах, кроме одного, вы можете:

1. Вычислить суммарный таргет по всем объектам.
2. Вычесть таргет конкретного объекта и разделить результат на `n_objects - 1`. 

Заметим, что пункт `1.` следует сделать для всех объектов. Также заметим, что пункт `2.` может быть реализован без циклов `for`.

Здесь может оказаться полезной функция .transform из **Способа 2** из примера.

In [35]:
all_data['sum_target'] = all_data.groupby('item_id').target.transform('sum')

all_data['n_objects'] = all_data.groupby('item_id').target.transform('count')

all_data['item_target_enc'] = (all_data['sum_target'] - all_data['target']) / (all_data['n_objects'] - 1)

all_data['item_target_enc'].fillna(0.3343, inplace=True) 


encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.480384831129305


Ожидаемый ответ 0.4803

### Smoothing (**1.25 балла**)

Необходимо реализовать smoothing с $\alpha = 100$. Используйте формулу:

$\frac{mean(target) \cdot nrows + globalmean \cdot \alpha }{nrows + \alpha}$,

где $globalmean=0.3343$. Заметим, что `nrows` - это количество объектов, принадлежащих конктертной категории, а не количество строк в датасете.

In [36]:
globalmean=0.3343
alpha = 100

all_data['mean_target'] = all_data.groupby('item_id').target.transform('mean')

all_data['n_rows'] = all_data.groupby('item_id').target.transform('count')

all_data['item_target_enc'] = ((all_data['mean_target'] * all_data['n_rows'] + globalmean * alpha)/
                               all_data['n_rows'] + alpha)

all_data['item_target_enc'].fillna(0.3343, inplace=True) 


encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])



0.48230046813303057


Ожидаемый ответ 0.4818

### Expanding mean схема (**1.25 балла**)

Необходимо реализовать *expanding mean* схему. Ее суть заключается в том, чтобы пройти по отсортированному в определенном порядке датасету (датасет сортируется в самом начале задания) и для подсчета счетчика для строки $m$ использовать строки от $0$ до $m-1$. Вам будет необходимо воспользоваться pandas функциями [`cumsum`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.DataFrameGroupBy.cumsum.html) и [`cumcount`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.cumcount.html).

cumsum - Cumulative sum for each group.

cumcount - Number each item in each group from 0 to the length of that group - 1.

In [37]:
all_data['sum_target'] = (all_data.groupby('item_id').target.transform('cumsum') - all_data['target'])
all_data['n_items'] = all_data.groupby('item_id').target.transform('cumcount')

all_data['item_target_enc'] = (all_data['sum_target'] / all_data['n_items'])

all_data['item_target_enc'].fillna(0.3343, inplace=True) 


encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.5025245211081701


Ожидаемый ответ 0.5025